### PR TITLE
BUGFIX - 0.1.3.82 - fix schedule next run not update bug #82

### PR DESCRIPTION
**Pull request by scattm**
- Models.Schedule unit test
- Fix bug on Schedule.update_next_run function. #82

### CODE REVIEW

File: DanceCat/Models.py [hunk 1]
```diff
@@ -542,43 +544,47 @@ def update_next_run(self, validated=False, frequency=60):
...
+        next_check = datetime.datetime.now() + relativedelta(seconds=interval)
+        if next_check <= self.next_run:
+            return
+
+        next_run_time = datetime.datetime.now().replace(second=0)
 
         if self.schedule_type == Constants.SCHEDULE_HOURLY:
-            next_run_time += relativedelta(
-                minute=self.minute_of_hour, seconds=-1
+            next_run_time = next_run_time + relativedelta(
```
Comment (caohhung) [inline DanceCat/Models.py]: += please

File: tests/test_model_schedule.py [hunk 2]
```diff
@@ -0,0 +1,604 @@
...
+from DanceCat import Models
+from DanceCat import Constants
+import pytest
+
+
+# Set fixed time at 2016-09-01 18:19:20.
+FIXED_TIME = datetime.datetime(2016, 9, 1, 18, 19, 20)
+
+
+@pytest.fixture
```
Comment (caohhung) [inline tests/test_model_schedule.py]: All fixture should be defined in https://github.com/scattm/DanceCat/blob/dev/tests/conftest.py

File: tests/test_model_schedule.py [hunk 3]
```diff
@@ -0,0 +1,604 @@
...
+def freeze_datetime(monkeypatch):
+    """Patch datetime.now function to return fixed timestamp."""
+    import datetime
```
Comment (caohhung) [inline tests/test_model_schedule.py]: duplicated import?

File: tests/test_model_schedule.py [hunk 4]
```diff
@@ -0,0 +1,604 @@
...
+    original_datetime = datetime.datetime
+
+    class FDateTimeMeta(type):
```
Comment (caohhung) [inline tests/test_model_schedule.py]: what does 'F' stand for?

File: tests/test_model_schedule.py [hunk 5]
```diff
@@ -0,0 +1,604 @@
...
+            return cls.frozen_time
+
+    monkeypatch.setattr(datetime, 'datetime', FDateTime)
+    FDateTime.freeze(original_datetime.now())
+    return FDateTime
+
+
+@pytest.fixture
+def setup_to_add_job(app, user_email):
+    """Setup data so we can test Schedule Model."""
```
Comment (caohhung) [inline tests/test_model_schedule.py]: Look like it missing 'db.session.remove()' huh?

File: tests/test_model_schedule.py [hunk 6]
```diff
@@ -0,0 +1,604 @@
...
+    job = Models.QueryDataJob(
+        'test job',
+        'select * from table_1',
+        user.user_id
+    )
+    db.session.add(job)
+    db.session.commit()
+
+
+def some_datetime_func():
```
Comment (caohhung) [inline tests/test_model_schedule.py]: https://github.com/scattm/DanceCat/blob/dev/tests/conftest.py please

File: tests/test_model_schedule.py [hunk 7]
```diff
@@ -0,0 +1,604 @@
...
+        'is_active': True
+    }
+
+    in_active_schedule_skeleton = {
+        'job_id': 1,
+        'user_id': 1,
+        'is_active': False
+    }
+
+    def test_should_add_new_schedules(self,
```
Comment (caohhung) [inline tests/test_model_schedule.py]: Please split this test case into smaller test cases
 + hourly_schedule 
 + daily_schedule

File: tests/test_model_schedule.py [hunk 8]
```diff
@@ -0,0 +1,604 @@
...
+        )
+        db.session.add(hourly_schedule)
+        db.session.commit()
+        datetime_now = datetime.datetime.now()
+        datetime_delta_created_on = datetime_now - hourly_schedule.created_on
+        datetime_delta_last_updated = \
+            datetime_now - hourly_schedule.last_updated
+
+        assert hourly_schedule.schedule_id == 1
+        assert hourly_schedule._is_active
```
Comment (caohhung) [inline tests/test_model_schedule.py]: who do you set it as private variable?
If You want to public it, please use public variable or write a method to cover it

File: tests/test_model_schedule.py (hunk 7 above)
Comment (caohhung) [inline tests/test_model_schedule.py]: We should discuss more about those test cases...If you test schedule without verifying it afterward, we should use 'real' time

File: tests/test_model_schedule.py [hunk 9]
```diff
@@ -0,0 +1,604 @@
...
+        assert not daily_schedule.is_active
+        assert daily_schedule.next_run > self.start_time_ugly
+        assert daily_schedule.minute_of_hour == 21
+        assert daily_schedule.hour_of_day == 13
+        # 0 for Monday -> 5 for Saturday
+        assert daily_schedule.day_of_week == 5
+        assert daily_schedule.day_of_month == 3
+        assert daily_schedule.schedule_type == Constants.SCHEDULE_DAILY
+
+        with pytest.raises(TypeError):
```
Comment (caohhung) [inline tests/test_model_schedule.py]: Should be a separated test case

File: tests/test_model_schedule.py [hunk 10]
```diff
@@ -0,0 +1,604 @@
...
+        )
+        db.session.add(schedule)
+        db.session.commit()
+
+        no_active_schedules = Models.Schedule.query.filter(
+            Models.Schedule.is_active
+        ).count()
+        assert no_active_schedules == 1
+
+        schedule.is_deleted = True
```
Comment (caohhung) [inline tests/test_model_schedule.py]: It isn't a good idea to access to an instance variable like this...Can we cover it by a method?

File: tests/test_model_schedule.py [hunk 11]
```diff
@@ -0,0 +1,604 @@
...
+                                   setup_to_add_job
+                                   ):
+        """Check if the schedule is valid or not."""
+        schedule = Models.Schedule(
+            schedule_type=Constants.SCHEDULE_HOURLY,
+            start_time=self.start_time_beautifully,
+            **self.active_schedule_skeleton
+        )
+
+        # Once upon a time
```
Comment (caohhung) [inline tests/test_model_schedule.py]: Sorry, this remind me http://www.imdb.com/title/tt0285823/

File: tests/test_model_schedule.py [hunk 12]
```diff
@@ -0,0 +1,604 @@
...
+        schedule.schedule_type = Constants.SCHEDULE_ONCE
+        with pytest.raises(ValueError):
+            schedule.update_start_time(self.start_time_ugly)
+
+        # Hourly
+        schedule.schedule_type = Constants.SCHEDULE_HOURLY
+        for moh in [
```
Comment (caohhung) [inline tests/test_model_schedule.py]: It's ok for this 'for loop' in the same line

 for moh in [-1, 60, 61]:

File: tests/test_model_schedule.py [hunk 13]
```diff
@@ -0,0 +1,604 @@
...
+        ).count()
+        assert no_active_schedules == 0
+
+        no_schedules = Models.Schedule.query.count()
+        assert no_schedules == 1
+
+    def test_should_schedule_valid(self,
+                                   setup_to_add_job
+                                   ):
+        """Check if the schedule is valid or not."""
```
Comment (caohhung) [inline tests/test_model_schedule.py]: Split into test cases please.

File: tests/test_model_schedule.py [hunk 14]
```diff
@@ -0,0 +1,604 @@
...
+        schedule = Models.Schedule(
```
Comment (caohhung) [inline tests/test_model_schedule.py]: If you're afraid this variable will be duplicated in each test case, please define it in class scope (class variable) with some specific name then reuse in each test case

File: tests/test_model_schedule.py [hunk 15]
```diff
@@ -0,0 +1,604 @@
...
+            -1, 60, 61
+        ]:
+            schedule.minute_of_hour = moh
+            assert not schedule.validate()
+        for moh in [
+            0, 4, 59
+        ]:
+            schedule.minute_of_hour = moh
+            assert schedule.validate()
+        # Do not care about others
```
Comment (caohhung) [inline tests/test_model_schedule.py]: Agreed ~

Comment (caohhung) [inline tests/test_model_schedule.py]: Missing a test case: minute_of_hour with default value?

File: tests/test_model_schedule.py [hunk 16]
```diff
@@ -0,0 +1,604 @@
...
+        schedule.hour_of_day = 99
+        schedule.day_of_week = 9
+        schedule.day_of_month = -4
+        assert schedule.validate()
+
+        # Daily
+        schedule.schedule_type = Constants.SCHEDULE_DAILY
+        for moh, hod in [
```
Comment (caohhung) [inline tests/test_model_schedule.py]: what does `hod` stand for? `hodor` ?

File: tests/test_model_schedule.py [hunk 17]
```diff
@@ -0,0 +1,604 @@
...
+        ]:
+            schedule.minute_of_hour = moh
+            schedule.hour_of_day = hod
+            assert schedule.validate()
+        # Do not care about others
+        schedule.day_of_week = 300
+        schedule.day_of_month = 8
+        assert schedule.validate()
+
+        # Weekly, 0 for Mon -> 6 for Sun
```
Comment (caohhung) [inline tests/test_model_schedule.py]: https://github.com/scattm/DanceCat/pull/91/files#diff-3f7ffe4f2d8d586cc575a3d678e0a198R154
Comment isn't consistent  with https://github.com/scattm/DanceCat/pull/91/files#diff-3f7ffe4f2d8d586cc575a3d678e0a198R154

File: tests/test_model_schedule.py [hunk 18]
```diff
@@ -0,0 +1,604 @@
...
+        ]:
+            schedule.minute_of_hour = moh
+            schedule.hour_of_day = hod
+            schedule.day_of_month = dom
+            assert schedule.validate()
+        # Do not care about others
+        schedule.day_of_week = 33
+        assert schedule.validate()
+
+    def test_should_add_schedule_once_exactly_time(
```
Comment (caohhung) [inline tests/test_model_schedule.py]: If it runs only once, should we call it 'schedule_once' or not?

File: DanceCat/Models.py (hunk 1 above)
Comment (scattm) [inline DanceCat/Models.py]: ops, forgot to switch back after working around

File: tests/test_model_schedule.py (hunk 5 above)
Comment (scattm) [inline tests/test_model_schedule.py]: It will call `app` fixture before doing any action

File: tests/test_model_schedule.py (hunk 2 above)
Comment (scattm) [inline tests/test_model_schedule.py]: Ok, forgot to put it back there lol

File: tests/test_model_schedule.py (hunk 4 above)
Comment (scattm) [inline tests/test_model_schedule.py]: Frozen...
Anyway switched to FrozenDateTimeMeta.

File: tests/test_model_schedule.py (hunk 6 above)
Comment (scattm) [inline tests/test_model_schedule.py]: this function use for test the `FrozenDateTime` mock, therefor should be in separate file. Will add on next commit.

File: tests/test_model_schedule.py (hunk 7 above)
Comment (scattm) [inline tests/test_model_schedule.py]: I don't "freeze" time in all the test, just for some test that need to ensure the result. Anyway agree with splitting the test case.

File: tests/test_model_schedule.py (hunk 8 above)
Comment (scattm) [inline tests/test_model_schedule.py]: Because it will be used, `Schedule.is_active` is combined of `Schedule._is_active` and `Schedule.is_deleted`.
In this test case I want ensure that `is_active` column in database is not `False` while `is_active` variable is False.

File: tests/test_model_schedule.py (hunk 10 above)
Comment (scattm) [inline tests/test_model_schedule.py]: We could use decorator, but I think it is ok for now since SQLAlchemy will cover with one more logic layer, and we also have one more layer at DB models (If it is not SQLite). For later when there are more logic that need to be covered, we will use method or decorator.

File: tests/test_model_schedule.py (hunk 8 above)
Comment (caohhung) [inline tests/test_model_schedule.py]: s.t is not right

File: tests/test_model_schedule.py (hunk 10 above)
Comment (caohhung) [inline tests/test_model_schedule.py]: ACK, pls kindly ignore this commend

File: tests/test_model_schedule.py (hunk 8 above)
Comment (scattm) [inline tests/test_model_schedule.py]: I don't think directly access to private variable is totally wrong. Thus I won't use it in normal case, I have decorators for this. This is just a test case.

File: tests/test_model_schedule.py (hunk 10 above)
Comment (scattm) [inline tests/test_model_schedule.py]: comment lolz

File: tests/test_model_schedule.py (hunk 15 above)
Comment (scattm) [inline tests/test_model_schedule.py]: Normally it will be automatically set on object initialization when you pass a start timestamp, or when you call update_start_time for later use. We should not set up things like this case, this test use to check if things will work in case s.o try violate the class variables.

File: tests/test_model_schedule.py (hunk 17 above)
Comment (scattm) [inline tests/test_model_schedule.py]: Why?

File: tests/test_model_schedule.py (hunk 18 above)
Comment (scattm) [inline tests/test_model_schedule.py]: No, even if will run only once, appoint s.th in the future will be call schedule.

File: tests/test_model_schedule.py (hunk 12 above)
Comment (scattm) [inline tests/test_model_schedule.py]: minute of hour FYI, I think it's easy to understand, isn't it?

File: tests/test_model_schedule.py (hunk 14 above)
Comment (scattm) [inline tests/test_model_schedule.py]: It's is not the same, these schedule have different type and values. That is why I use skeleton dicts for it.

File: tests/test_model_schedule.py (hunk 16 above)
Comment (scattm) [inline tests/test_model_schedule.py]: Hour of day. Same as above.

File: tests/test_model_schedule.py (hunk 7 above)
Comment (scattm) [inline tests/test_model_schedule.py]: As for adding, I don't see the necessary to split to smaller test case. Just want to add, deeper logic will be in other test case.

File: tests/test_model_schedule.py (hunk 9 above)
Comment (scattm) [inline tests/test_model_schedule.py]: Done

File: tests/test_model_schedule.py (hunk 13 above)
Comment (scattm) [inline tests/test_model_schedule.py]: Done

**Comment by scattm** [comment]
Updated, when you are done let me know then I will squash

File: tests/test_model_schedule.py [hunk 19]
```diff
@@ -156,54 +81,56 @@ def test_should_add_new_schedules(self,
...
-        db.session.add(schedule)
+        db.session.add(schedule_will_delete)
         db.session.commit()
 
         no_active_schedules = Models.Schedule.query.filter(
             Models.Schedule.is_active
         ).count()
         assert no_active_schedules == 1
 
-        schedule.is_deleted = True
```
Comment (caohhung) [inline tests/test_model_schedule.py]: why did you rename this variable? It's doesn't make sense if you separated into test cases and each case has it own name/description

File: tests/test_model_schedule.py [hunk 20]
```diff
@@ -213,110 +140,132 @@ def test_would_delete_schedule(self,
...
+        schedule_once = Models.Schedule(
             schedule_type=Constants.SCHEDULE_HOURLY,
             start_time=self.start_time_beautifully,
             **self.active_schedule_skeleton
         )
 
         # Once upon a time
-        schedule.schedule_type = Constants.SCHEDULE_ONCE
+        schedule_once.schedule_type = Constants.SCHEDULE_ONCE
         with pytest.raises(ValueError):
```
Comment (caohhung) [inline tests/test_model_schedule.py]: why could a valid schedule raise `ValueError` ? Or I missed something?

File: tests/test_model_schedule.py [hunk 21]
```diff
@@ -213,110 +140,132 @@ def test_would_delete_schedule(self,
...
+            assert not schedule_hourly.validate()
+        for minute_of_hour in [0, 4, 59]:
+            schedule_hourly.minute_of_hour = minute_of_hour
+            assert schedule_hourly.validate()
         # Do not care about others
-        schedule.hour_of_day = 99
-        schedule.day_of_week = 9
-        schedule.day_of_month = -4
-        assert schedule.validate()
+        schedule_hourly.hour_of_day = 99
```
Comment (caohhung) [inline tests/test_model_schedule.py]: same above :)

File: tests/test_model_schedule.py [hunk 22]
```diff
@@ -0,0 +1,553 @@
...
+            datetime.datetime(2016, 9, 3, 13, 22, 00),
+            datetime.datetime(2016, 9, 3, 13, 22, 34),
+        ]:
+            freeze_datetime.freeze(self.start_time_beautifully)
+            schedule.update_start_time(self.start_time_ugly)
+            freeze_datetime.freeze(cur_time)
+            schedule.update_next_run(interval=29)
+            assert schedule.next_run == expected_next_run
+
+    def test_schedule_daily_update_next_run(
```
Comment (caohhung) [inline tests/test_model_schedule.py]: Should split into test cases

**Comment by caohhung** [comment]
Everything else is good

File: tests/test_model_schedule.py (hunk 20 above)
Comment (caohhung) [inline tests/test_model_schedule.py]: so, rename method's name?

Comment (scattm) [inline tests/test_model_schedule.py]: Why? If they pass wrong value, I raise ValueError, nothing wrong...

File: tests/test_model_schedule.py (hunk 19 above)
Comment (scattm) [inline tests/test_model_schedule.py]: I agree about variable names, but I don't think it does not make sense lol

File: tests/test_model_schedule.py (hunk 22 above)
Comment (scattm) [inline tests/test_model_schedule.py]: In those test we have a lot of duplicate code. Let's it be for now, we should fix it at Model level first.

**Comment by caohhung** [comment]
Please squash all fixup commits
